Third example (#224)
ZIB-IOL/FrankWolfe.jl: docs/src/examples.md

                push!(missing_ratings, (I[idx], J[idx]))
            else
                push!(present_ratings, (I[idx], J[idx]))
            end
        end
    end
end

function f(X)
    r = 0.0
    for (i, j) in present_ratings
        r += 0.5 * (X[i, j] - rating_matrix[i, j])^2
    end
    return r
end

function grad!(storage, X)
    storage .= 0
    for (i, j) in present_ratings
        storage[i, j] = X[i, j] - rating_matrix[i, j]
    end
    return nothing
end

function test_loss(X)
    r = 0.0
    for (i, j) in missing_ratings
        r += 0.5 * (X[i, j] - rating_matrix[i, j])^2
    end
    return r
end

function project_nuclear_norm_ball(X; radius=1.0)
    U, sing_val, Vt = svd(X)
    if (sum(sing_val) <= radius)
        return X, -norm_estimation * U[:, 1] * Vt[:, 1]'
    end
    sing_val = FrankWolfe.projection_simplex_sort(sing_val, s=radius)
    return U * Diagonal(sing_val) * Vt', -norm_estimation * U[:, 1] * Vt[:, 1]'
end

norm_estimation = 10 * Arpack.svds(rating_matrix, nsv=1, ritzvec=false)[1].S[1]

const lmo = FrankWolfe.NuclearNormLMO(norm_estimation)
const x0 = FrankWolfe.compute_extreme_point(lmo, ones(size(rating_matrix)))
const k = 100

> FrankWolfe.benchmark_oracles(
>     f,
>     (str, x) -> grad!(str, x),
>     () -> randn(size(rating_matrix)),
>     lmo;
>     k=100,
> )

Reviewer: ```suggestion
```

gradient = spzeros(size(x0)...)
gradient_aux = spzeros(size(x0)...)

> function build_callback(trajectory_arr)

Reviewer: ```suggestion
# Define a custom callback computing at each iteration the test loss
function build_callback(trajectory_arr)
```

    xscalelog=[:log, :identity, :log, :identity],
    legend_position=[:bottomleft, nothing, nothing, nothing],
)
plot!(size=(3000, 2000), legendfontsize=30, annotationfontsize=30, guidefontsize=30, tickfontsize=30)
```

## Matrix Completion

> We present another example that is about matrix completion. The idea is, given a partially observed matrix ``Y\in\mathbb{R}{m\times n}``, to find

Reviewer: ```suggestion
We present another example that is about matrix completion. The idea is, given a partially observed matrix ``Y\in\mathbb{R}^{m\times n}``, to find
```

> ``X\in\mathbb{R}{m\times n}`` to minimize the sum of squared errors from the observed entries while 'completing' the matrix ``Y``, i.e. filling the unobserved

Reviewer: ```suggestion
``X\in\mathbb{R}^{m\times n}`` to minimize the sum of squared errors from the observed entries while 'completing' the matrix ``Y``, i.e. filling the unobserved
```

    xscalelog=[:log, :identity, :log, :identity],
    legend_position=[:bottomleft, nothing, nothing, nothing],
)
plot!(size=(3000, 2000), legendfontsize=30, annotationfontsize=30, guidefontsize=30, tickfontsize=30)
```

## Matrix Completion

We present another example that is about matrix completion. The idea is, given a partially observed matrix ``Y\in\mathbb{R}{m\times n}``, to find
``X\in\mathbb{R}{m\times n}`` to minimize the sum of squared errors from the observed entries while 'completing' the matrix ``Y``, i.e. filling the unobserved

> entries to match ``Y`` as good as possible. Again, a detailed explanation can be found in chapter 4.2 of

Reviewer: ```suggestion
entries to match ``Y`` as good as possible. A detailed explanation can be found in section 4.2 of
```
